src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParser.cs

```diff
@@ -0,0 +1,573 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Routing.Patterns
+{
+    internal static class RoutePatternParser
+    {
+        private const char Separator = '/';
+        private const char OpenBrace = '{';
+        private const char CloseBrace = '}';
+        private const char EqualsSign = '=';
+        private const char QuestionMark = '?';
+        private const char Asterisk = '*';
+        private const string PeriodString = ".";
+
+        internal static readonly char[] InvalidParameterNameChars = new char[]
+        {
+            Separator,
+            OpenBrace,
+            CloseBrace,
+            QuestionMark,
+            Asterisk
+        };
+
+        public static RoutePattern Parse(string pattern)
+        {
+            if (pattern == null)
+            {
+                throw new ArgumentNullException(nameof(pattern));
+            }
+
+            var trimmedPattern = TrimPrefix(pattern);
+
+            var context = new TemplateParserContext(trimmedPattern);
+            var segments = new List<RoutePatternPathSegment>();
+
+            while (context.MoveNext())
+            {
+                var i = context.Index;
+
+                if (context.Current == Separator)
+                {
+                    // If we get here is means that there's a consecutive '/' character.
+                    // Templates don't start with a '/' and parsing a segment consumes the separator.
+                    throw new RoutePatternException(pattern, Resources.TemplateRoute_CannotHaveConsecutiveSeparators);
+                }
+
+                if (!ParseSegment(context, segments))
+                {
+                    throw new RoutePatternException(pattern, context.Error);
+                }
+
+                // A successful parse should always result in us being at the end or at a separator.
+                Debug.Assert(context.AtEnd() || context.Current == Separator);
+
+                if (context.Index <= i)
+                {
+                    throw new InvalidProgramException("Infinite loop in the parser. This is a bug.");
```

Comment:
I don't like "This is a bug." in error message to user

Reply:
What would you suggest? This is not a message a user should be able to see. I want this code to crash hard if there's an infinite loop.

Reply:
I would have something like "An unexpected infinite loop detected in the parser." and then a comment saying we should never reach here